application-ideas-api/src/main/java/com/xwiki/ideas/internal/DefaultIdeasManager.java

```diff
@@ -0,0 +1,189 @@
+import org.xwiki.model.reference.SpaceReference;
+import org.xwiki.velocity.tools.JSONTool;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.user.api.XWikiUser;
+import com.xwiki.ideas.IdeasException;
+import com.xwiki.ideas.IdeasManager;
+
+/**
+ * @version $Id$
+ * @since 1.4
+ */
+@Component
+@Singleton
+public class DefaultIdeasManager implements IdeasManager
+{
+    static final LocalDocumentReference IDEA_CLASS_REFERENCE = new LocalDocumentReference("Ideas", "IdeasClass");
+
+    static final String VOTERS_AGAINST_KEY = "against";
+
+    static final String NUMBER_OF_AGAINST_VOTES_KEY = "nbagainst";
+
+    static final String VOTERS_FOR_KEY = "supporters";
+
+    static final String NUMBER_OF_FOR_VOTES_KEY = "nbvotes";
+
+    static final String COMMA = ",";
+
+    private static final Map<String, String> SUPP_TO_NR_KEY = new HashMap<>();
+
+    static {
+        SUPP_TO_NR_KEY.put(VOTERS_AGAINST_KEY, NUMBER_OF_AGAINST_VOTES_KEY);
+        SUPP_TO_NR_KEY.put(VOTERS_FOR_KEY, NUMBER_OF_FOR_VOTES_KEY);
+    }
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Override public String getRestUrl(DocumentReference documentReference)
+    {
+        String contextPath = contextProvider.get().getRequest().getContextPath();
+        StringBuilder stringBuilder = new StringBuilder();
+        try {
+            stringBuilder.append(contextPath);
+            stringBuilder.append("/rest/wikis/");
+            stringBuilder
+                .append(URLEncoder.encode(documentReference.getWikiReference().getName(), XWiki.DEFAULT_ENCODING));
+            for (SpaceReference spaceReference : documentReference.getSpaceReferences()) {
+                stringBuilder.append("/spaces/");
+                stringBuilder.append(URLEncoder.encode(spaceReference.getName(), XWiki.DEFAULT_ENCODING));
+            }
+            stringBuilder.append("/pages/");
+            stringBuilder.append(URLEncoder.encode(documentReference.getName(), XWiki.DEFAULT_ENCODING));
+            stringBuilder.append("/idea");
+            return stringBuilder.toString();
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(String.format("Failed to retrieve the REST URL of the document: [%s]",
+                documentReference), e);
+        }
+    }
+
+    @Override public String vote(DocumentReference documentReference, String voteType) throws IdeasException
+    {
+        XWikiContext xcontext = contextProvider.get();
+        XWiki xWiki = xcontext.getWiki();
+        Map<String, Object> result = new HashMap<>();
+        try {
+            XWikiDocument mydoc = xWiki.getDocument(documentReference, xcontext);
+            BaseObject ideasObj = mydoc.getXObject(IDEA_CLASS_REFERENCE);
+            XWikiUser user = xWiki.checkAuth(xcontext);
```

Comment:
```suggestion
            DocumentReference userReference = xcontext.getUserReference();
```

The Ideas Manager shouldn't care about authentication or user rights, because the Ideas Manager is not available directly to the user. It is exposed through a script service or a REST resource, so it's their responsibility to check the user rights (you usually don't put a lock on each door in your house, you put a lock only on the front door).